Source/GSStoryboardTransform.m

```diff
@@ -0,0 +1,866 @@
+#import <Foundation/NSXMLDocument.h>
+#import <Foundation/NSXMLNode.h>
+#import <Foundation/NSXMLElement.h>
+#import <Foundation/NSUUID.h>
+#import <Foundation/NSBundle.h>
+#import <Foundation/NSKeyedArchiver.h>
+
+#import "AppKit/NSSeguePerforming.h"
+#import "AppKit/NSStoryboard.h"
+#import "AppKit/NSStoryboardSegue.h"
+#import "AppKit/NSNibDeclarations.h"
+#import "AppKit/NSViewController.h"
+#import "AppKit/NSWindowController.h"
+
+#import "GSStoryboardTransform.h"
+#import "GSFastEnumeration.h"
+
+@interface NSStoryboardSegue (__private__)
+- (void) _setDestinationController: (id)controller; 
+- (void) _setSourceController: (id)controller;
+@end
+
+@interface NSStoryboardSegue (__StoryboardPrivate__)
+// Private to this class...
+- (void) _setKind: (NSString *)k;
+- (void) _setRelationship: (NSString *)r;
+- (NSString *) _kind;
+- (NSString *) _relationship;
+@end
+
+// this needs to be set on segues
+@implementation NSStoryboardSegue (__StoryboardPrivate__)
+- (void) _setKind: (NSString *)k
+{
+  ASSIGN(_kind, k);
+}
+
+- (void) _setRelationship: (NSString *)r
+{
+  ASSIGN(_relationship, r);
+}
+
+- (NSString *) _kind
+{
+  return _kind;
+}
+
+- (NSString *) _relationship
+{
+  return _relationship;
+}
+@end
+
+@implementation NSStoryboardSeguePerformAction
+- (id) target
+{
+  return _target;
+}
+
+- (void) setTarget: (id)target
+{
+  ASSIGN(_target, target);
+}
+
+- (SEL) action
+{
+  return _action;
+}
+
+- (void) setAction: (SEL)action
+{
+  _action = action;
+}
+
+- (NSString *) selector
+{
+  return NSStringFromSelector(_action);
+}
+
+- (void) setSelector: (NSString *)s
+{
+  _action = NSSelectorFromString(s);
+}
+
+- (id) sender
+{
+  return _sender;
+}
+
+- (void) setSender: (id)sender
+{
+  ASSIGN(_sender, sender);
+}
+
+- (NSString *) identifier
+{
+  return _identifier;
+}
+
+- (void) setIdentifier: (NSString *)identifier
+{
+  ASSIGN(_identifier, identifier);
+}
+
+- (NSString *) kind
+{
+  return _kind;
+}
+
+- (void) setKind: (NSString *)kind
+{
+  ASSIGN(_kind, kind);
+}
+
+- (NSDictionary *) identifierToSegueMap
+{
+  return _identifierToSegueMap;
+}
+
+- (void) setIdentifierToSegueMap: (NSDictionary *)table
+{
+  ASSIGN(_identifierToSegueMap, table);
+}
+
+- (NSStoryboard *) storyboard
+{
+  return _storyboard;
+}
+
+- (void) setStoryboard: (NSStoryboard *)storyboard
+{
+  ASSIGN(_storyboard, storyboard);
+}
+
+- (id) nibInstantiate
+{
+  return self;
+}
+
+- (void) dealloc
+{
+  RELEASE(_storyboard);
+  RELEASE(_identifierToSegueMap);
+  RELEASE(_kind);
+  RELEASE(_identifier);
+  RELEASE(_sender);
+  [super dealloc];
+}
+
+- (IBAction) doAction: (id)sender
+{
+  if (_sender != nil)
+    {
+      [_sender performSegueWithIdentifier: _identifier
+                                   sender: _sender];
+    }
+  else // This is a special case where there is no source controller and we don't ask "should"
+    {
+      NSMapTable *mapTable = [_identifierToSegueMap objectForKey: @"application"];
```

Comment:
I will try to explain in more detail the changes I am suggesting. (But I don't have the time at the moment to do so for each place, sorry.)
 For example here this is the only place where the NSStoryboardSeguePerformAction uses the _identifierToSegueMap. You could try to store the NSStoryboardSegue directly in the NSStoryboardSeguePerformAction in the method instantiateControllerWithIdentifier:creator:. That way the internal complexity of the code would be greatly reduced. 
Of course this is only possible if the object get created  in the correct order. But I think this is possible.

Reply:
One of the reasons I did it the way it is currently done is to delay the processing of the scenes until they are needed.   I think I understand what you're saying, but I would ask that we defer this change to a later date.    The changes, as they are, work very well and are well tested.  I will take a look into what you're talking about.

Reply:
I think this is something BETTER done in a later iteration.  The current code works.

Reply:
I'm deferring this to a later commit.  I really want to be done with this code.

Reply:
The issue with storing NSStoryboardSegue this way is that the NSStoryboardSegueAction isn't unarchived until it's loaded in instantiateControllerWithIdentifier: method in NSStoryboard.  This is why I had to organize the code in the way I did by mapping the segue itself to the Segue identifiers associated with a given controller.   So we have a map of controllers, then we have a map of segues which belong to each of those controllers.  When the controller is instantiated (based on it's sceneId) the XML I created for the NSStoryboardSegueAction is then unarchived.   

Either way we will need to unarchive the scene/document in order to story the segue on the action which is what we are pretty much doing anyway.

Reply:
I will go ahead and make this change and see if it simplifies the code.  Since we have everything we need in the instantiateControllerWithIdentifier:... method.